scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/synchronization/Synchronization.java

```diff
@@ -0,0 +1,1135 @@
+     * The first parameter receives the key, and the second parameter receives the binding associated with this key.
+     *
+     * For example, to append a {@code String} msg to a value when the value will be defined
+     * mapping:
+     *
+     * <pre> {@code
+     * store.waitUntilThen(channel, key, "{ k, v -> v != null}", "{k, v -> v.concat(\"" + msg + "\")}")}</pre>
+     *
+     * <p>If the function returns {@code null}, the mapping is removed (or
+     * remains absent if initially absent).  If the predicate or function itself throw an
+     * (unchecked) exception, the exception is rethrown as a ClosureEvaluationException, and the current mapping
+     * is left unchanged.
+     *
+     * The default implementation is equivalent to performing the following
+     * steps for this {@code map}, then returning the current value or
+     * {@code null} if absent:
+     *
+     * <pre> {@code
+     * V oldValue = map.get(key);
+     * while (!predicate.test(key, oldValue)) {
+     *  oldValue = map.get(key)
+     * }
+     * map.compute(key, thenRemappingFunction);
+     * }</pre>
+     *
+     * <p>This implementation guarantees that the operation is performed atomically on the channel map, which means that as soon as another operation modifies the underlying structure to meet the predicate,
+     * the remappingFunction will be executed at once and atomically, without letting another operation to interfere.
+     *
+     * @param channel channel identifier
+     * @param key key with which the specified value is to be associated
+     * @param thenRemappingFunction the function to compute a value
+     * @return a {@link PredicateActionResult} object which contains the result of the predicate evaluation and the new value associated with the key after applying thenRemappingFunction
+     * @throws NullPointerException if the specified key is null and
+     *         this map does not support null keys, or the
+     *         remappingFunction is null
+     * @throws CompilationException if an error occurred during the compilation of the specified action
+     * @throws ClosureEvaluationException if an error occurred during the evaluation of the specified action
+     * @throws InvalidChannelException if the channel does not exist
+     * @throws IOException if an error occurred when persisting the modified channel on disk
+     * @since 1.8
+     * @see java.util.Map#compute(Object, BiFunction)
+     * @see java.util.function.BiPredicate
+     * @see java.util.function.BiFunction
+     * @see <a href="http://groovy-lang.org/closures.html">Groovy Closures</a>
+     */
+    PredicateActionResult waitUntilThen(String channel, String key, String predicate, String thenRemappingFunction)
+            throws InvalidChannelException, CompilationException, IOException;
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)